Experiment with error handling

AbbiePBC opened the pull request:
this is still v much a WIP (as is the main branch), but idk how to error handle properly, pls give suggestions

abbie-currington committed:
experiement w error handling

abbie-currington committed:
add back in error handling for local environment and crowding environment being swapped, such that all errors are reported together

jose-mut-lopez commented on src/main.rs:
```diff
@@ -1,4 +1,7 @@
 use std::ops::AddAssign;
+use std::error;
+use std::fmt;
```
asdf

jose-mut-lopez commented on src/main.rs:
```diff
@@ -19,23 +23,62 @@ struct Flock {
     time_per_frame: i32,
 }
 
+fn validate_distances(max_dist_before_boid_is_crowded: f32, max_dist_of_local_boid: f32) -> Result<(), CreationError> {
```
shouldn't max_dist_before_boid_is_crowded be min_dist_before_boid_is_crowded?

abbie-currington replied:
ifl `max_dist_before_boid_is_no_longer_crowded` is what I was trying to say, which makes more sense to me, when considering x,y increasing away from the boid rather than towards, will rename to that

jose-mut-lopez commented on src/main.rs:
```diff
@@ -158,12 +201,49 @@ impl AddAssign for Boid {
 
 
+#[derive(PartialEq, Debug)]
+struct IntegerBetweenZeroAndOne(f32, String);
```
if it's a float it's not an integer XD

jose-mut-lopez committed:
simplify add_assign

jose-mut-lopez committed:
make assertion failures more readable

if this test ever fails, the previous assertion would say "hey,
I expected a list of 3 elements but there were 4", so to know even
which one is missing you need to debug with a breakpoint or put logs.

You know this already, this is just a reminder to make assertions
useful when they trigger

jose-mut-lopez commented on src/main.rs:
```diff
@@ -9,6 +12,7 @@ fn main() {
 
 }
```
as a general suggestion, don't do " function+test, function+test, function+test, use all functions in main". do "test, function1, main with function 1, test, function2, add function2 in main".

In other words, work in vertical slices so that you have a working program more often, even if incomplete. as example, look at my commits https://github.com/jmmut/boids/commits/master. you can see some commits like:
- basic bird drawing
- you can steer a bird
- you can accelerate or brake a bird
- spawn bot birds
- decent first version of boids
- add personal space. birds move away if they are too close
- add cohesion (go to the center of neighbours)

and I can tell you the whole program was runnable in each of those commits, and it did something, but was only complete on the last commit.

(Don't look at the code, it's an ugly motherfucker 🙈 I was just having fun)

jose-mut-lopez commented on src/main.rs:
```diff
@@ -19,23 +23,62 @@ struct Flock {
+        return Err(errors);
+    }
+
+    return Ok((optional_factors[0].unwrap(), optional_factors[1].unwrap(), optional_factors[2].unwrap()));
```
I can see how you are imitating what we did in the amservice. I don't love this approach but don't have much better suggestions. I added my take on the PR (#2) I added to your PR.

jose-mut-lopez committed:
rework the Result usage

jose-mut-lopez committed:
remove IntegerBetweenZeroAndOne

It wasn't really used as a struct. It would maybe make sense to make
Flock have "some_factor: IntegerBetweenZeroAndOne,", but then having
the factor name inside IntegerBetweenZeroAndOne is confusing, so
you would want something like:

```
struct IntegerBetweenZeroAndOne(f32);

impl IntegerBetweenZeroAndOne {
    fn new(value: f32, factor: String) -> Result<IntegerBetweenZeroAndOne, CreationError> {
        match value {
            x if x < 0.0 => Err(CreationError::FactorShouldBeMoreThanZero(factor)),
            x if x > 1.0 => Err(CreationError::FactorShouldBeLessThanOne(factor)),
            x => Ok(Self (value))
        }
    }
}
```
and then something like:
```
let repulsion = IntegerBetweenZeroAndOne::new(...)
if err {
    accumulate error
}
...
if no errors {
    self.repulsion = repulsion.unwrap()
}
```
or some variation of that, but it felt too complicated for the benefit.

jose-mut-lopez committed:
allow checking the errors after calling Flock::new

Note how the main still can convert to anyhow::Error

abbie-currington committed:
:memo: - update teaching comments from Jose

abbie-currington committed:
:recycle: - rename variable
